TypeCobol.Analysis/Graph/CfgDotFileForNodeGenerator.cs

```diff
@@ -0,0 +1,128 @@
+        /// <param name="block">The block to get the dot format name.</param>
+        /// <returns>The block's name</returns>
+        protected override string BlockName(BasicBlock<Node, D> block)
+        {
+            string name = base.BlockName(block);
+            ControlFlowGraphBuilder<D>.BasicBlockForNode nodeBlock = (ControlFlowGraphBuilder<D>.BasicBlockForNode)block;
+            if (!block.HasFlag(BasicBlock<Node, D>.Flags.Start) && nodeBlock.Tag != null)
+            {
+                string tag = nodeBlock.Tag.ToString();
+                if (!tag.Equals(ControlFlowGraphBuilder<D>.ROOT_SECTION_NAME))
+                {
+                    name = tag.ToUpper() + ". " + name;
+                }
+            }
+            return name;
+        }
+
+        /// <summary>
+        /// Memorized emitted group to avoid infinite recursion.
+        /// </summary>
+        internal HashSet<int> EmittedGroupIndices
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// Emit a basic block, with also handling recursive BasicBlock Groups that can appear for
+        /// instance for recursive PERFORM.
+        /// </summary>
+        /// <param name="block">The BasicBlock to emit</param>
+        /// <param name="cfg">The target Control Flow Graph that contains the Basic Block</param>
+        /// <returns>true</returns>
+        protected override bool EmitBasicBlock(BasicBlock<Node, D> block, ControlFlowGraph<Node, D> cfg)
+        {
+            bool bResult = base.EmitBasicBlock(block, cfg);
+            if ((block is ControlFlowGraphBuilder<D>.BasicBlockForNodeGroup) && ! block.HasFlag(BasicBlock<Node, D>.Flags.GroupGrafted))
+            {
+                ControlFlowGraphBuilder<D>.BasicBlockForNodeGroup group = (ControlFlowGraphBuilder<D>.BasicBlockForNodeGroup)block;
+                if (EmittedGroupIndices == null)
+                {
+                    EmittedGroupIndices = new HashSet<int>();
+                }
+                StringWriter sw = new StringWriter();
+                if (!EmittedGroupIndices.Contains(group.GroupIndex))
+                {
+                    EmittedGroupIndices.Add(group.GroupIndex);
+                    //we are emitting a sub graph.
+                    sw.WriteLine("subgraph cluster_" + group.GroupIndex + '{');
+                    sw.WriteLine("color = blue;");
+                    if (group.Group.Count > 0)
+                    {
+                        sw.WriteLine(string.Format("label = \"{0}\";", ((ControlFlowGraphBuilder<D>.BasicBlockForNode)group.Group.First.Value).Tag));
+                        CfgDotFileForNodeGenerator<D> cfgDot = new CfgDotFileForNodeGenerator<D>(cfg);
+                        cfgDot.EmittedGroupIndices = EmittedGroupIndices;
+                        cfgDot.FullInstruction = this.FullInstruction;
+                        cfgDot.Writer = sw;
+                        cfgDot.DigraphBuilder = new StringBuilder();
```

Comment:
Why creating a StringBuilder if we already have a TextWriter ? It would be more efficient to continue writing subgraph on the writer.

Reply:
We are creating a Dot SubGraph for BasicBlockForNodeGroup, to be merge din the main the parent Dot gaph. The Writer is reused (cfgDot.Writer = sw;), but the instruction for directed edges are written at tail construction of the graph in cfgDot.Writer = sw;).